File: src/main/java/com/joecollins/graphics/screens/generic/MapBuilder.java
package com.joecollins.graphics.screens.generic;

import com.joecollins.bindings.Binding;
import com.joecollins.bindings.BindingReceiver;
import com.joecollins.graphics.components.MapFrame;
import com.joecollins.graphics.components.MapFrameBuilder;
import com.joecollins.graphics.utils.ColorUtils;
import com.joecollins.models.general.Party;
import java.awt.Color;
import java.awt.Shape;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.apache.commons.lang3.tuple.Pair;

public class MapBuilder {

  private final BindingReceiver<List<Pair<Shape, Color>>> winners;
  private final BindingReceiver<List<Shape>> mapFocus;
  private final BindingReceiver<String> mapHeader;

  public <T> MapBuilder(
      Binding<Map<T, Shape>> shapes,
      Binding<Map<T, Result>> winners,
      Binding<List<T>> focus,
      Binding<String> headerBinding) {
    BindingReceiver<Map<T, Shape>> shapesReceiver = new BindingReceiver<>(shapes);
    Binding<List<ImmutablePair<Shape, Result>>> shapesToParties =
        shapesReceiver
            .getBinding()
            .merge(
                winners,
                (s, w) ->
                    s.entrySet().stream()
                        .map(
                            e -> {
                              Result winnerParty = w.get(e.getKey());
                              return ImmutablePair.of(e.getValue(), winnerParty);
                            })
                        .collect(Collectors.toList()));
    this.mapFocus =
        new BindingReceiver<>(shapesReceiver.getBinding().merge(focus, this::createFocusShapes));
    this.winners =
        new BindingReceiver<>(
            shapesToParties.merge(
                mapFocus.getBinding(),
                (r, f) ->
                    r.stream()
                        .map(p -> ImmutablePair.of(p.left, extractColor(f, p.left, p.right)))
                        .collect(Collectors.toList())));
    this.mapHeader = new BindingReceiver<>(headerBinding);
  }

  public <T> MapBuilder(
      Binding<Map<T, Shape>> shapes,
      Binding<T> selectedShape,
      Binding<Result> leadingParty,
      Binding<List<T>> focus,
      Binding<String> header) {
    this(shapes, selectedShape, leadingParty, focus, () -> null, header);
  }

  public <T> MapBuilder(
      Binding<Map<T, Shape>> shapes,
      Binding<T> selectedShape,
      Binding<Result> leadingParty,
      Binding<List<T>> focus,
      Binding<List<T>> additionalHighlight,
      Binding<String> header) {
    BindingReceiver<Map<T, Shape>> shapesReceiver = new BindingReceiver<>(shapes);
    Binding<ImmutablePair<T, Result>> leaderWithShape =
        selectedShape.merge(leadingParty, ImmutablePair::new);
    mapFocus =
        new BindingReceiver<>(shapesReceiver.getBinding().merge(focus, this::createFocusShapes));
    Binding<List<Shape>> additionalFocusShapes =
        shapesReceiver.getBinding().merge(additionalHighlight, this::createFocusShapes);
    mapHeader = new BindingReceiver<>(header);
    Binding<List<Pair<Shape, Color>>> shapeWinners =
        shapesReceiver
            .getBinding()
            .merge(
                leaderWithShape,
                (shp, ldr) ->
                    shp.entrySet().stream()
                        .map(
                            e -> {
                              Color color;
                              if (e.getKey().equals(ldr.left)) {
                                color =
                                    Optional.ofNullable(ldr.right)
                                        .map(Result::getColor)
                                        .orElse(Color.BLACK);
                              } else {
                                color = Color.LIGHT_GRAY;
                              }
                              return ImmutablePair.of(e.getValue(), color);
                            })
                        .collect(Collectors.toList()));
    Binding<List<Shape>> allFocusShapes =
        mapFocus
            .getBinding()
            .merge(
                additionalFocusShapes,
                (l1, l2) -> {
                  if (l1 == null) return l2;
                  if (l2 == null) return l1;
                  return Stream.concat(l1.stream(), l2.stream())
                      .distinct()
                      .collect(Collectors.toList());
                });
    Binding<List<Pair<Shape, Color>>> focusedShapeWinners =
        shapeWinners.merge(
            allFocusShapes,
            (sw, f) -> {
              if (f == null) {
                return sw;
              }
              return sw.stream()
                  .map(
                      e -> {
                        if (f.contains(e.getLeft())) {
                          return e;
                        }
                        return ImmutablePair.of(e.getLeft(), new Color(220, 220, 220));
                      })
                  .collect(Collectors.toList());
            });
    winners = new BindingReceiver<>(focusedShapeWinners);
  }

  private <T> List<Shape> createFocusShapes(Map<T, Shape> shapes, List<T> focus) {
    return focus == null
        ? null
        : focus.stream().filter(shapes::containsKey).map(shapes::get).collect(Collectors.toList());
  }

  public MapFrame createMapFrame() {
    if (mapHeader == null) {
      return null;
    }
    return MapFrameBuilder.from(winners.getBinding())
        .withFocus(mapFocus.getBinding())
        .withHeader(mapHeader.getBinding())
        .build();
  }

  private static Color extractColor(List<Shape> focus, Shape shape, Result winner) {
    if (winner != null) {
      return winner.getColor();
    } else if (focus == null || focus.isEmpty() || focus.contains(shape)) {
      return Color.LIGHT_GRAY;
    } else {
      return new Color(220, 220, 220);
    }
  }

  public static class Result {

    private final Party party;
    private final boolean elected;

    public Result(Party party, boolean elected) {
      this.party = party;
      this.elected = elected;
    }

    public static Result elected(Party party) {
      return new Result(party, true);
    }

    public static Result leading(Party party) {
      return new Result(party, false);
    }

    public Color getColor() {
      if (party == null) {
        return Color.BLACK;
      }
      if (elected) {
        return party.getColor();
      }
      return ColorUtils.lighten(party.getColor());
    }
  }
}

File: src/main/java/com/joecollins/graphics/screens/generic/MapBuilder.kt
package com.joecollins.graphics.screens.generic

import com.joecollins.bindings.Binding
import com.joecollins.bindings.Binding.Companion.fixedBinding
import com.joecollins.bindings.BindingReceiver
import com.joecollins.graphics.components.MapFrame
import com.joecollins.graphics.components.MapFrameBuilder
import com.joecollins.models.general.PartyResult
import java.awt.Color
import java.awt.Shape

class MapBuilder<T> {
    private val winners: BindingReceiver<List<Pair<Shape, Color>>>
    private val mapFocus: BindingReceiver<List<Shape>?>
    private val mapHeader: BindingReceiver<String?>
    private var notes: Binding<String?>? = null

    constructor(
        shapes: Binding<Map<T, Shape>>,
        winners: Binding<Map<T, PartyResult?>>,
        focus: Binding<List<T>?>,
        headerBinding: Binding<String?>
    ) : this(shapes, winners, Pair(focus, fixedBinding(null)), headerBinding)

    constructor(
        shapes: Binding<Map<T, Shape>>,
        winners: Binding<Map<T, PartyResult?>>,
        focus: Pair<Binding<List<T>?>, Binding<List<T>?>>,
        headerBinding: Binding<String?>
    ) {
        val shapesReceiver: BindingReceiver<Map<T, Shape>> = BindingReceiver(shapes)
        val shapesToParties = shapesReceiver
                .getBinding()
                .merge(winners) { s, w ->
                    s.entries
                            .map {
                                val winnerParty = w[it.key]
                                Pair(it.value, winnerParty)
                            }
                            .toList()
                }
        mapFocus = BindingReceiver(shapesReceiver.getBinding().merge(focus.first) { shp, foc -> createFocusShapes(shp, foc) })
        val additionalFocus = BindingReceiver(shapesReceiver.getBinding().merge(focus.second) { shp, foc -> createFocusShapes(shp, foc) })
        val allFocusShapes = mapFocus.getBinding().merge(additionalFocus.getBinding()) { a, b ->
            when {
                a == null -> b
                b == null -> a
                else -> listOf(a, b).flatten()
            }
        }
        this.winners = BindingReceiver(
                shapesToParties.merge(allFocusShapes) { r: List<Pair<Shape, PartyResult?>>, f: List<Shape>? ->
                    r.map { Pair(it.first, extractColor(f, it.first, it.second)) }
                            .toList()
                })
        mapHeader = BindingReceiver(headerBinding)
    }

    constructor(
        shapes: Binding<Map<T, Shape>>,
        selectedShape: Binding<T>,
        leadingParty: Binding<PartyResult?>,
        focus: Binding<List<T>?>,
        header: Binding<String?>
    ) : this(shapes, selectedShape, leadingParty, focus, Binding.fixedBinding(null), header)

    constructor(
        shapes: Binding<Map<T, Shape>>,
        selectedShape: Binding<T>,
        leadingParty: Binding<PartyResult?>,
        focus: Binding<List<T>?>,
        additionalHighlight: Binding<List<T>?>,
        header: Binding<String?>
    ) {
        val shapesReceiver: BindingReceiver<Map<T, Shape>> = BindingReceiver(shapes)
        val leaderWithShape: Binding<Pair<T, PartyResult?>> = selectedShape.merge(leadingParty) { left, right -> Pair(left, right) }
        mapFocus = BindingReceiver(shapesReceiver.getBinding().merge(focus) { shp, foc -> createFocusShapes(shp, foc) })
        val additionalFocusShapes = shapesReceiver.getBinding().merge(additionalHighlight) { shp, foc -> createFocusShapes(shp, foc) }
        mapHeader = BindingReceiver(header)
        val shapeWinners: Binding<List<Pair<Shape, Color>>> = shapesReceiver
                .getBinding()
                .merge(leaderWithShape) { shp, ldr ->
                    shp.entries
                            .map {
                                val color =
                                    if (it.key == ldr.first) {
                                        ldr.second?.color ?: Color.BLACK
                                    } else {
                                        Color.LIGHT_GRAY
                                    }
                                Pair(it.value, color)
                            }
                            .toList()
                }
        val allFocusShapes = mapFocus
                .getBinding()
                .merge(
                        additionalFocusShapes
                ) { l1: List<Shape>?, l2: List<Shape>? ->
                    when {
                        l1 == null -> l2
                        l2 == null -> l1
                        else -> listOf(l1, l2).flatten().distinct()
                    }
                }
        val focusedShapeWinners = shapeWinners.merge(allFocusShapes) { sw: List<Pair<Shape, Color>>, f: List<Shape>? ->
            if (f == null) {
                sw
            } else {
                sw.map {
                            if (f.contains(it.first)) {
                                it
                            } else {
                                Pair(it.first, Color(220, 220, 220))
                            }
                        }
                        .toList()
            }
        }
        winners = BindingReceiver(focusedShapeWinners)
    }

    private fun <T> createFocusShapes(shapes: Map<T, Shape>, focus: List<T>?): List<Shape>? {
        return focus
                ?.filter { key: T -> shapes.containsKey(key) }
                ?.map { key: T -> shapes[key]!! }
                ?.toList()
    }

    fun withNotes(notes: Binding<String?>): MapBuilder<T> {
        this.notes = notes
        return this
    }

    fun createMapFrame(): MapFrame {
        return MapFrameBuilder.from(winners.getBinding())
                .withFocus(mapFocus.getBinding())
                .withHeader(mapHeader.getBinding())
                .let { map -> notes?.let { n -> map.withNotes(n) } ?: map }
                .build()
    }

    companion object {
        private fun extractColor(focus: List<Shape>?, shape: Shape, winner: PartyResult?): Color {
            val isInFocus = focus == null || focus.isEmpty() || focus.contains(shape)
            return winner?.color?.takeIf { isInFocus }
                    ?: (if (isInFocus) {
                        Color.LIGHT_GRAY
                    } else {
                        Color(220, 220, 220)
                    })
        }
    }
}
